incubator/hnc/pkg/controllers/object_controller_new.go

```diff
@@ -38,25 +60,307 @@ type ObjectReconcilerNew struct {
 
 	// GVK is the group/version/kind handled by this reconciler.
 	GVK schema.GroupVersionKind
+
+	// Affected is a channel of event.GenericEvent (see "Watching Channels" in
+	// https://book-v1.book.kubebuilder.io/beyond_basics/controller_watches.html) that is used to
+	// enqueue additional objects that need updating.
+	Affected chan event.GenericEvent
 }
 
 // +kubebuilder:rbac:groups=*,resources=*,verbs=get;list;watch;create;update;patch;delete
 
+// SyncNamespace can be called manually by the HierarchyReconciler when the hierarchy changes.
+// It enqueues all the current objects in the namespace and local copies of the original objects
+// in the ancestors.
+func (r *ObjectReconcilerNew) SyncNamespace(ctx context.Context, log logr.Logger, ns string) error {
+	log = log.WithValues("gvk", r.GVK)
+	ul := &unstructured.UnstructuredList{}
+	ul.SetGroupVersionKind(r.GVK)
+	if err := r.List(ctx, ul, client.InNamespace(ns)); err != nil {
+		log.Error(err, "Couldn't list objects")
+		return err
+	}
+
+	// Enqueue all the current objects in the namespace because some of them may have been deleted.
+	for _, inst := range ul.Items {
+		if r.isExcluded(log, &inst) {
+			continue
+		}
+		// We don't need the entire canonical object here but only its metadata.
+		// Using canonical copy is the easiest way to get an object with its metadata set.
+		co := object.Canonical(&inst)
+		co.SetNamespace(inst.GetNamespace())
+		log.Info("Enqueuing existing object for reconciliation", "affected", co.GetName())
+		r.Affected <- event.GenericEvent{Meta: co}
```

Comment:
Maybe worth factoring this logic out into its own function? `enqueuePropagatedObjects` has similar logic, as does `enqeueDescendants`.

That also makes the comment more visible in multiple places too.